### ISSUE TITLE
Usage with CTC loss function or multi-multi-class classification

### ISSUE THREAD

**Issue by cathoderaymission** [pytorch/captum#612]
I have a simple question which, looking through the source code, I am not sure _select_targets allows.

I am using the CTC loss function with a 1D convolutional network, so I get an output of (batchsize, X, Y) where **X** is increasingly large, and for each **x** there are **Y** possible words. So for example, I would have an output of (batchsize, 100, 10).

How can I format the list (or whatever necessary) to the "target" option of any function (eg attributegradients) to get the appropriate portion of the input related for each classification of **X**? So for example I want to get the attribution at (0, 0, 5) then (0, 1, 3), (0, 2, 9), etc.. I would then like a vector of the attributions from the input for each of these classifications (all 100) so I can see what part of the input is related to each classification.

Is this possible? Thanks!

**Comment by vivekmig** [comment]
Hi @cathoderaymission , it seems like what you are trying to do is obtain attributions for multiple targets for the same input example simultaneously. For most attribution methods (particularly gradient-based methods), this isn't directly possible, since attribution for each target requires separate input gradients, but the inputs are the same for each, rather than the typical case where a single target is chosen for each input example within the batch. There are a few ways you can get around this:
1. Compute each attribution separately. In you example, you can call attribute once for each, with the target first being (0,5), then (1, 3), then (2, 9) and so on.
2. Expand the input batch with the same input example repeated K times, where K is the number of attributions you want. Input batch size would then be K, and target could then be provided as something like [(0,5), (1,3), (2,9), ...] corresponding to output indices (0, 0, 5), (1, 1, 3), (2, 2, 9), ...
3. If you instead want to obtain an aggregated attribution for all the outputs together (e.g. which inputs affect the sum of all the desired outputs), this can be accomplished with a wrapper function as follows, which can be provided as the model when instantiating an attribution method, and no target would be necessary:
```
def wrapper_forward(inp):
     out = model(inp) 
     return out[:,0,5] + out[:,1,3] + out[:,2,9] + ....
```
Note that this approach wouldn't allow identifying the attributions for each output separately.

The above approaches should work with any of the attribution methods. For a few methods, particularly perturbation-based methods such as feature ablation and feature permutation, we can compute exactly the attributions you are looking for in an easier way. The reason is that with perturbation-based methods, we can observe the change in multiple output targets without repeating the input. This could be done with a wrapper method for the model as follows:
```
 def wrapper_perturbation_forward(inp):
        out = model(inp) 
        return torch.stack((out[0,0,5], out[0,1,3], out[0,2,9], ...)
```

Hope this helps!

**Comment by craymichael** [comment]
Hi, this is covered by the FAQ target section. For outputs with more than two dimensions, pass a tuple target such as `(time_index, class_index)`, repeat the input if you want multiple separate attributions, or wrap the forward function to aggregate the selected outputs into one scalar. Closing as documented. https://captum.ai/docs/faq
